tests/e2e_tests/flatfield_e2e.py

```diff
@@ -0,0 +1,368 @@
+
+    # Nonlinearity calibration
+    nonlin_dat = np.genfromtxt(nonlin_path, delimiter=",")
+    nonlinear_cal = data.NonLinearityCalibration(nonlin_dat, pri_hdr=pri_hdr, ext_hdr=ext_hdr,
+                                                input_dataset=mock_input_dataset)
+    nonlinear_cal.save(filedir=flat_outputdir, filename="mock_nonlinearcal.fits" )
+    this_caldb.create_entry(nonlinear_cal)
+
+    # KGain
+    kgain_val = 8.7
+    kgain = data.KGain(np.array([[kgain_val]]), pri_hdr=pri_hdr, ext_hdr=ext_hdr, 
+                    input_dataset=mock_input_dataset)
+    kgain.save(filedir=flat_outputdir, filename="mock_kgain.fits")
+    this_caldb.create_entry(kgain)
+
+    # NoiseMap
+    with fits.open(fpn_path) as hdulist:
+        fpn_dat = hdulist[0].data
+    with fits.open(cic_path) as hdulist:
+        cic_dat = hdulist[0].data
+    with fits.open(dark_path) as hdulist:
+        dark_dat = hdulist[0].data
+    noise_map_dat_img = np.array([fpn_dat, cic_dat, dark_dat])
+    noise_map_dat = np.zeros((3, detector.detector_areas['SCI']['frame_rows'], detector.detector_areas['SCI']['frame_cols']))
+    rows, cols, r0c0 = detector.unpack_geom('SCI', 'image')
+    noise_map_dat[:, r0c0[0]:r0c0[0]+rows, r0c0[1]:r0c0[1]+cols] = noise_map_dat_img
+    noise_map_noise = np.zeros([1,] + list(noise_map_dat.shape))
+    noise_map_dq = np.zeros(noise_map_dat.shape, dtype=int)
+    err_hdr = fits.Header()
+    err_hdr['BUNIT'] = 'detected EM electrons'
+    ext_hdr['B_O'] = 0
+    ext_hdr['B_O_ERR'] = 0
+    noise_map = data.DetectorNoiseMaps(noise_map_dat, pri_hdr=pri_hdr, ext_hdr=ext_hdr,
+                                    input_dataset=mock_input_dataset, err=noise_map_noise,
+                                    dq = noise_map_dq, err_hdr=err_hdr)
+    noise_map.save(filedir=flat_outputdir, filename="mock_detnoisemaps.fits")
+    this_caldb.create_entry(noise_map)
+
+    ####### Run the walker on some test_data
+
+    recipe = walker.autogen_recipe(l1_flatfield_filelist, flat_outputdir)
+     ### Modify they keywords of some of the steps
+    for step in recipe['steps']:
+        # if step['name'] in ["desmear", "cti_correction"]:
+        #     step['skip'] = True
+        if step['name'] == "create_onsky_flatfield":
+            step['keywords']['n_pix'] = 165 # full shaped pupil FOV
+    walker.run_recipe(recipe, save_recipe_file=True)
+
+
+    ####### Test the flat field result
+    flat_filename = l1_flatfield_filelist[0].split(os.path.sep)[-1].replace("_L1_", "_L2a_")[:-5] + "_flatfield.fits"
+    flat = data.FlatField(os.path.join(flat_outputdir, flat_filename))
+    good_region = np.where(flat.data != 1)
+    diff = flat.data - input_flat # compute residual from true
+    smoothed_diff = scipy.ndimage.gaussian_filter(diff, 1.4) # smooth by the size of the resolution element, since we care about that
+    print(np.std(smoothed_diff[good_region]))
+    assert np.std(smoothed_diff[good_region]) < 0.0071
```

Comment:
presumably this number is a requirement somewhere? we could reference it?

Reply:
Added, I just pulled it from the Zellem+ calibration paper